src/metabase/api/permission_graph.clj

```diff
@@ -7,105 +7,110 @@
    [clojure.spec.alpha :as s]
    [clojure.spec.gen.alpha :as gen]
    [clojure.walk :as walk]
-   [metabase.util :as u]))
+   [malli.core :as mc]
+   [malli.error :as me] ;; umd/describe
+   [malli.util :as mut]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [trs]]))
 
 (defmulti ^:private convert
   "convert values from the naively converted json to what we REALLY WANT"
   first)
 
-(defmethod convert :kw->int
-  [[_ k]]
-  (Integer/parseInt (name k)))
-
-(defmethod convert :str->kw
-  [[_ s]]
-  (keyword s))
+(defmethod convert :kw->int [[_ k]] (Integer/parseInt (name k)))
+(defmethod convert :str->kw [[_ s]] (keyword s))
 
 ;; Convert a keyword to string without excluding the namespace.
 ;; e.g: :schema/name => "schema/name".
 ;; Primarily used for schema-name since schema are allowed to have "/"
 ;; and calling (name s) returning a substring after "/".
-(defmethod convert :kw->str
-  [[_ s]]
-  (u/qualified-name s))
-
-(defmethod convert :nil->none
-  [[_ _]]
-  :none)
-
-(defmethod convert :identity
-  [[_ x]]
-  x)
+(defmethod convert :kw->str [[_ s]] (u/qualified-name s))
+(defmethod convert :nil->none [[_ _]] :none)
+(defmethod convert :identity [[_ x]] x)
+(defmethod convert :global-execute [[_ x]] x)
+(defmethod convert :db-exeute [[_ x]] x)
 
-(defmethod convert :global-execute
-  [[_ x]]
-  x)
+;;; --------------------------------------------------- Common ----------------------------------------------------
 
-(defmethod convert :db-exeute
-  [[_ x]]
-  x)
+;;integer schema that knows how to decode itself from the :123 sort of shape used in perm-graphs
+(def decodable-kw-int
+  [:int {:decode/perm-graph
+         (fn thing [kw-int]
+           (if (int? kw-int)
+             kw-int
+             (Integer/parseInt (name kw-int))))}])
```

Comment:
This replaces the creative use of `convert` above. Once the other permission graph types are supplanted we can remove `convert`.